climt/_core/initialization.py

```diff
@@ -428,11 +454,11 @@ def init_interface_level_tau_longwave(array_dims, quantity_description, initial_
         'default_value': 'sea',
         'dtype': 'a100'
     },
-    'snow_ice_temperature': {
+    'snow_ice_temperature_poly': {
         'dims': ['x', 'y'],
         'units': 'dimensionless',
         'default_value': CubicSpline(
-            np.linspace(0, 50, 50), 270.*np.ones(50)),
+            np.linspace(0, 50, 50), 260.*np.ones(50)),
```

Comment:
Is there a reason to do this as a CubicSpline of a single value, instead of a lambda function that returns `260. * np.ones_like(x)`?

Reply:
This component can also take care of snow or ice melting or growing. If you use a fixed vertical resolution for the snow/ice pack, then the size of the array has to grow or shrink. To get around this, the temperature profile is stored as a spline object, and is evaluated at whatever discrete points needed by the component.

Reply:
There are a couple problems with this:
- Currently, the way snow_ice_temperature_poly is used, you only ever have a constant value for the snow+ice temperature. It is initialized with a CubicSpline that has a constant profile, and always returns a CubicSpline with a constant profile in the diagnostics of IceSheet. This tells me that it should just use a constant value instead.
- The second problem is that this breaks the Sympl API around units. The units on this quantity are currently set to 'dimensionless', even though they should be Kelvin. If you did set the units, then unit conversions would not work, since CubicSpline objects can't be unit converted.
- It's also not immediately clear what the CubicSpline takes as input (it takes grid indices).
- There is also a bug here - when you have a growing layer let's say of size N last timestep, now size N+1, [this line](https://github.com/CliMT/climt/blob/develop/climt/_components/surface_ice.py#L215) will try to get the temperature at cell N+1 from a CubicSpline that only has values set from 0 to N. It will do this by cubic extrapolation, which is probably not what you want, since it uses the same cubic function at point N in the spline as at point N-1. None of this is clear from looking at the default value or at the _properties dictionary.

I think the way to do it that's compatible with Sympl would be either:
- If you really do only need constant temperatures, just use a single value.
- If you need non-constant temperatures, you can have DataArrays whose dimension lengths change from timestep to timestep. I think this would require some modifications to Sympl (to allow a dimension to change size), but that would be perfectly reasonable. This is something that we should make Sympl compatible with, anyways.

Reply:
Ah, I see that it actually does use non-constant temperatures (mistook `new_temp`, which should probably be `new_temperature` for a scalar).

Reply:
It shouldn't have units, since it is a polynomial, which cannot be evaluated as you point out. There is a separate output field which contains the gridded output (called `snow_ice_temperature`, will rename!).

While it is feasible to change dimension sizes, I don't see it as a typical use case, and moreover will cause issues if we want to write to netcdf. I decided on this approach taking these issues into consideration.

Reply:
The CubicSpline does have units, though, because the values it outputs have units. The way it's set up right now, the properties don't document what snow_ice_temperature_poly is supposed to be, what it takes in, or the units of its output. That goes against a lot of our goals for writing a clear, self-documenting component.

You said that issues when trying to write to netCDF is a reason to use the CubicSpline, but that doesn't work because you'll still get issues writing a state containing an array of CubicSpline objects to netCDF, won't you? I don't think you can actually save them to a netCDF file.

It's not a "typical" use case, but it *is* a use case that can come up in a variety of contexts (e.g. mesh refinement).

Is there a reason other than writing to netCDF why you would use CubicSpline over an array that has a changing dimension size?

Reply:
Yes, I have been worried about that as well -- how do we document inputs and outputs? This particular component is simple, but there is currently no standard way of documenting *_properties.

Might be worth opening an issue in the sympl repo to discuss this in a more focused way.

No, we cannot write CubicSpline to a netCDF file, but there is an actual output field `snow_ice_temperature` that has a fixed shape that _can_ be written. This field is created by evaluating the spline at a fixed number of levels, which is decided during component instantiation.

For this particular component, using a CubicSpline makes the actual algorithm we implement to calculate ice temperatures and growth/melting much easier. We would otherwise need an array whose dimension varies in space (since you want to use the same dz for all columns), and I don't think I have the bandwidth to carefully think that out.

I agree there is a definite use case for variable shaped arrays, but I think we need a concrete component whose demands sympl will need to satisfy to take up further design towards this end. Until then, I think our energy is better spent elsewhere -- a lot of work still remains to make sympl/climt actually fully usable in terms of convenience.

Reply:
I do agree that having an array whose dimension varies in space (or whose z dimension is the max z dimension size, with missing values for shorter columns) is undesirable.

Perhaps you should add a "description" property for this input. It's not something standardized in Sympl, but Sympl explicitly allows you to include additional keys that aren't required in *_properties.

I think you should label the units with the units of the output. Even though unit conversions won't work (and will give strange errors), you have exactly the same problem with "dimensionless" (which is actually a unit, it's just the unitless unit). Since probably no other component will be using snow_and_ice_temperature_poly, using "degK" shouldn't cause code problems, and should make the units a bit clearer. If other components do need to interact with this, they'll just have to do so without converting units (i.e. also requiring Kelvin).

What do you think about the problems of polynomial extrapolation in the case of growing layers?

Reply:
yes, and I'm thinking of adding a `describe()` method to components which gives a nice summary of everything -- much like `head()` for pandas dataframes.

Will do the units thing. They way it is written, it never has to extrapolate. The increase in height is done at the end of the routine. This creates a small error in the surface temperature, but it should remain small unless a huge amount of precipitation happens, dumping a lot of snow. Will check it carefully again, though.